pkg/serviceimport/map.go

```diff
@@ -2,21 +2,53 @@ package serviceimport
 
 import (
 	"sync"
+	"sync/atomic"
 
 	lighthousev2a1 "github.com/submariner-io/lighthouse/pkg/apis/lighthouse.submariner.io/v2alpha1"
 )
 
+type clusterInfo struct {
+	Ip     string
+	Name   string
+	Weight uint64
+}
+
 type serviceInfo struct {
 	key         string
 	clusterInfo map[string]string
 	ipList      []string
+	queue       []clusterInfo // clusters queue [each could have an 1 vertual ip or an ip pool of nodes (headless)]
+	rrCount     uint64
 }
 
 type Map struct {
 	svcMap map[string]*serviceInfo
 	sync.RWMutex
 }
 
+func (m *Map) GetQueue(namespace, name string) ([]clusterInfo, bool) {
+	m.RLock()
+	defer m.RUnlock()
+
+	if val, ok := m.svcMap[keyFunc(namespace, name)]; ok {
+		return val.queue, len(val.queue) > 0
+	}
+
+	return nil, false
+}
+
+func (m *Map) GetAndIncRRCounter(namespace, name string) (uint64, bool) {
+	m.RLock()
+	defer m.RUnlock()
+
+	if val, ok := m.svcMap[keyFunc(namespace, name)]; ok {
+		atomic.AddUint64(&val.rrCount, 1)
+		return val.rrCount, true
+	}
+
+	return 0, false
+}
```

Comment:
I think it would be ideal to not expose the RR implementation to callers. I suggest to define a function to select an IP  using a passed function to check clusters:

```
    func (m *Map) SelectIP(namespace, name string, checkCluster func(string) bool) (string, bool) {
	queue, counter := func() ([]clusterInfo, *uint64) {
		m.RLock()
		defer m.RUnlock()

		si, ok := m.svcMap[keyFunc(namespace, name)]
		if !ok {
			return nil, nil
		}

		return si.rrQueue, &si.rrCount
	}()

	queueLength := len(queue)
	for i := 0; i < queueLength; i++ {
		c := atomic.LoadUint64(counter)
		atomic.AddUint64(counter, 1)

		info := queue[c%uint64(queueLength)]
		if checkCluster(info.Name) {
			return info.Ip, true
		}
	}

	return "", false
    }


```

Reply:
BTW - 2 points:
1) Eventually, I think we would need to get the loadbalancing logic out of the Map object
2) This will require to change several test cases... see the new commits (: